### PR TITLE
add merge token

### PR DESCRIPTION
**Pull request by zkJoaquin**
# What ❔

## Checklist

- [ ] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.

### CODE REVIEW

File: l2-contracts/contracts/bridge/L2ERC20Bridge.sol [hunk 1]
```diff
@@ -116,6 +161,25 @@ contract L2ERC20Bridge is IL2Bridge, Initializable {
         emit WithdrawalInitiated(msg.sender, _l1Receiver, _l2Token, _amount);
     }
 
+    /// @notice Initiates a withdrawal by burning merge funds on the contract and sending the message to L1
+    /// where tokens would be unlocked
+    /// @param _l1Receiver The account address that should receive funds on L1
+    /// @param _l2Token The L2 token address which is withdrawn
+    /// @param _amount The total amount of tokens to be withdrawn
+    function withdrawFromMerge(address _l1Receiver, address _l2Token, uint256 _amount) external {
+        IERC20Upgradeable(_l2Token).safeTransferFrom(msg.sender, address(this), _amount);
```
Comment (zkbenny) [inline l2-contracts/contracts/bridge/L2ERC20Bridge.sol:170]: get _l2MergeToken from portal
IERC20Upgradeable(_l2MergeToken).safeTransferFrom(msg.sender, address(this), _amount);

File: l2-contracts/contracts/bridge/L2ERC20Bridge.sol [hunk 2]
```diff
@@ -116,6 +161,25 @@ contract L2ERC20Bridge is IL2Bridge, Initializable {
...
+        IMergeTokenPortal(MERGE_TOKEN_PORTAL).withdraw(_l2Token, _amount, msg.sender);
+        IL2StandardToken(_l2Token).bridgeBurn(msg.sender, _amount);
```
Comment (zkbenny) [inline l2-contracts/contracts/bridge/L2ERC20Bridge.sol:172]: IL2StandardToken(_l2Token).bridgeBurn(address(this), _amount);

File: l2-contracts/contracts/bridge/L2ERC20Bridge.sol [hunk 3]
```diff
@@ -116,6 +161,25 @@ contract L2ERC20Bridge is IL2Bridge, Initializable {
...
     }
 
+    /// @notice Initiates a withdrawal by burning merge funds on the contract and sending the message to L1
+    /// where tokens would be unlocked
+    /// @param _l1Receiver The account address that should receive funds on L1
+    /// @param _l2Token The L2 token address which is withdrawn
+    /// @param _amount The total amount of tokens to be withdrawn
+    function withdrawFromMerge(address _l1Receiver, address _l2Token, uint256 _amount) external {
+        IERC20Upgradeable(_l2Token).safeTransferFrom(msg.sender, address(this), _amount);
+        IMergeTokenPortal(MERGE_TOKEN_PORTAL).withdraw(_l2Token, _amount, msg.sender);
```
Comment (zkbenny) [inline l2-contracts/contracts/bridge/L2ERC20Bridge.sol:171]: IMergeTokenPortal(MERGE_TOKEN_PORTAL).withdraw(_l2Token, _amount, address(this));

File: l2-contracts/contracts/bridge/L2ERC20Bridge.sol [hunk 4]
```diff
@@ -89,6 +99,41 @@ contract L2ERC20Bridge is IL2Bridge, Initializable {
...
+            l1TokenAddress[expectedL2Token] = _l1Token;
+        } else {
+            require(currentL1Token == _l1Token, "gg"); // Double check that the expected value equal to real one
+        }
+
+        IL2StandardToken(expectedL2Token).bridgeMint(address(this), _amount);
+        IERC20Upgradeable(expectedL2Token).safeApprove(MERGE_TOKEN_PORTAL, _amount);
+        IMergeTokenPortal(MERGE_TOKEN_PORTAL).deposit(expectedL2Token, _amount, _l2Receiver);
+
+        emit FinalizeDeposit(_l1Sender, _l2Receiver, expectedL2Token, _amount);
```
[the remark this answers is not in the document]
Comment (zkJoaquin) [reply l2-contracts/contracts/bridge/L2ERC20Bridge.sol]: add
`event FinalizeDepositToMerge(
        address indexed l1Sender,
        address indexed l2Receiver,
        address indexed l2Token,
        address mergeToken,
        uint256 amount
    );`

File: l1-contracts/contracts/bridge/L1ERC20Bridge.sol [hunk 5]
```diff
@@ -151,6 +151,28 @@ contract L1ERC20Bridge is IL1Bridge, IL1BridgeLegacy, ReentrancyGuard {
...
+    /// @dev If the token is bridged for the first time, the L2 token and merge token contract will be deployed. Note however, that the
+    /// newly-deployed token does not support any custom logic, i.e. rebase tokens' functionality is not supported.
+    /// @param _l2Receiver The account address that should receive funds on L2
+    /// @param _l1Token The L1 token address which is deposited
+    /// @param _amount The total amount of tokens to be bridged
+    /// @param _l2TxGasLimit The L2 gas limit to be used in the corresponding L2 transaction
+    /// @param _l2TxGasPerPubdataByte The gasPerPubdataByteLimit to be used in the corresponding L2 transaction
+    /// @return l2TxHash The L2 transaction hash of depositToMerge finalization
+    /// NOTE: the function doesn't use `nonreentrant` modifier, because the inner method does.
+    function depositToMerge(
```
Comment (zkbenny) [inline l1-contracts/contracts/bridge/L1ERC20Bridge.sol:166]: remove this legacy deposit

File: l1-contracts/contracts/bridge/L1ERC20Bridge.sol [hunk 6]
```diff
@@ -183,11 +205,74 @@ contract L1ERC20Bridge is IL1Bridge, IL1BridgeLegacy, ReentrancyGuard {
...
+    /// the funds would be lost.
+    /// @return l2TxHash The L2 transaction hash of depositToMerge finalization
+    function depositToMerge(
+        address _l2Receiver,
+        address _l1Token,
+        uint256 _amount,
+        uint256 _l2TxGasLimit,
+        uint256 _l2TxGasPerPubdataByte,
+        address _refundRecipient
+    ) public payable nonReentrant returns (bytes32 l2TxHash) {
```
Comment (zkbenny) [inline l1-contracts/contracts/bridge/L1ERC20Bridge.sol:250]: replace public with external

File: l2-contracts/contracts/bridge/L2ERC20Bridge.sol [hunk 7]
```diff
@@ -69,6 +79,47 @@ contract L2ERC20Bridge is IL2Bridge, Initializable {
...
+        bytes calldata _data
+    ) external payable {
+        address expectedL2Token = _getExpectedL2Token(_l1Token, _data);
+        IMergeTokenPortal.SourceTokenInfo memory sourceTokenInfo = MERGE_TOKEN_PORTAL.getSourceTokenInfos(
+            expectedL2Token
+        );
+        uint256 afterBalance = sourceTokenInfo.balance + _amount;
+
+        // Ensure that MergeToken can be deposit
+        if (sourceTokenInfo.isSupported && !sourceTokenInfo.isLocked && afterBalance <= sourceTokenInfo.depositLimit) {
```
Comment (zkbenny) [inline l2-contracts/contracts/bridge/L2ERC20Bridge.sol:109]: can we use try cache?

**Pull request merged**